pkg/operator/backupcontroller/backupcontroller.go

```diff
@@ -0,0 +1,364 @@
+		kubeClient:          kubeClient,
+		targetImagePullSpec: targetImagePullSpec,
+		featureGateAccessor: accessor,
+	}
+
+	syncer := health.NewDefaultCheckingSyncWrapper(c.sync)
+	livenessChecker.Add("BackupController", syncer)
+
+	return factory.New().ResyncEvery(1*time.Minute).
+		WithSync(syncer.Sync).ToController("BackupController", eventRecorder.WithComponentSuffix("backup-controller"))
+}
+
+func (c *BackupController) sync(ctx context.Context, _ factory.SyncContext) error {
+	if enabled, err := backuphelpers.AutoBackupFeatureGateEnabled(c.featureGateAccessor); !enabled {
+		if err != nil {
+			klog.V(4).Infof("BackupController error while checking feature flags: %v", err)
+		}
+		return nil
+	}
+
+	jobsClient := c.kubeClient.BatchV1().Jobs(operatorclient.TargetNamespace)
+	currentJobs, err := jobsClient.List(ctx, v1.ListOptions{LabelSelector: "app=" + backupLabel + ",state!=processed"})
+	if err != nil {
+		return fmt.Errorf("BackupController could not list backup jobs, error was: %w", err)
+	}
+
+	// we only allow to run one at a time, if there's currently a job running then we will skip it in this reconciliation step
+	runningJobs := findRunningJobs(currentJobs)
+	if len(runningJobs) > 0 {
+		klog.V(4).Infof("BackupController already found [%d] running jobs, skipping", len(runningJobs))
+		return nil
+	}
+
+	jobIndexed := indexJobsByBackupLabelName(currentJobs)
+	backupsClient := c.backupsClient.EtcdBackups()
+	backups, err := backupsClient.List(ctx, v1.ListOptions{LabelSelector: "state!=processed"})
+	if err != nil {
+		return fmt.Errorf("BackupController could not list etcdbackups CRDs, error was: %w", err)
+	}
+
+	if backups == nil {
+		return nil
+	}
+
+	var backupsToRun []operatorv1alpha1.EtcdBackup
+	for _, item := range backups.Items {
+		if backupJob, ok := jobIndexed[item.Name]; ok {
+			klog.V(4).Infof("BackupController backup job with name [%s] found, reconciling status", backupJob.Name)
+			err := reconcileJobStatus(ctx, jobsClient, backupsClient, backupJob, item)
+			if err != nil {
+				return fmt.Errorf("BackupController could not reconcile job status: %w", err)
+			}
+
+			delete(jobIndexed, item.Name)
+			continue
+		}
+
+		backupsToRun = append(backupsToRun, *item.DeepCopy())
+	}
+
+	if len(jobIndexed) > 0 {
+		klog.V(4).Infof("BackupController found dangling jobs without corresponding backup, removing")
+		for _, job := range jobIndexed {
+			err := jobsClient.Delete(ctx, job.Name, v1.DeleteOptions{})
+			if err != nil {
+				return fmt.Errorf("BackupController could not delete danging job [%s]: %w", job.Name, err)
+			}
+		}
+	}
+
+	if len(backupsToRun) == 0 {
+		klog.V(4).Infof("BackupController no backups to reconcile, skipping")
+		return nil
+	}
+
+	// in case of multiple backups requested, we're trying to reconcile in order of their names (also to reduce flakiness in tests)
+	sort.Slice(backupsToRun, func(i, j int) bool {
+		return strings.Compare(backupsToRun[i].Name, backupsToRun[j].Name) < 0
+	})
```

Comment:
For the CronJob we'll hopefully disallow this by setting a [forbid concurrency](https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/#concurrency-policy) policy but yes the user could still create multiple EtcdBackup requests at once or in quick succession.

Reply:
neither the cronjob nor the job doesn't really give you a guarantee that there's only ever a single pod running. The backup logic needs to be idempotent anyway.

Would you rather use Forbid or Replace as a strategy?

Reply:
we can continue this convo in the other PR, but I'll take Forbid for the time being.

Reply:
Yeah no I agree this needs to be idempotent. I just commenting on how we'd use the forbid strategy at the cronjob level on top of disallowing parallel runs at this controller's level (since there's no guarantee the user won't create multiple requests). So this is good as is.